document/json/decoder_test.go

```diff
@@ -0,0 +1,630 @@
+	*/
+	"json.Number to big.Float": {
+		json: []byte(strconv.FormatFloat(math.MaxFloat64, 'e', -1, 64)),
+		actual: func() interface{} {
+			var x big.Float
+			return &x
+		}(),
+		want: func() *big.Float {
+			return big.NewFloat(math.MaxFloat64)
+		}(),
+	},
+	"float64 to big.Float": {
+		json: []byte(strconv.FormatFloat(math.MaxFloat64, 'e', -1, 64)),
+		actual: func() interface{} {
+			var x big.Float
+			return &x
+		}(),
+		want: func() *big.Float {
+			return big.NewFloat(math.MaxFloat64)
+		}(),
+		disableJSONNumber: true,
+	},
+	"json.Number to big.Int": {
+		json: []byte(strconv.FormatInt(math.MaxInt64, 10)),
+		actual: func() interface{} {
+			var x big.Int
+			return &x
+		}(),
+		want: func() *big.Int {
+			return big.NewInt(math.MaxInt64)
+		}(),
+	},
+	"float64 to big.Int": {
+		json: []byte(strconv.FormatInt(math.MaxInt32, 10)),
+		actual: func() interface{} {
+			var x big.Int
+			return &x
+		}(),
+		want: func() *big.Int {
+			return big.NewInt(math.MaxInt32)
+		}(),
+		disableJSONNumber: true,
+	},
+}
+
+func MustJSONUnmarshal(v []byte, useJSONNumber bool) interface{} {
+	var jv interface{}
+	decoder := stdjson.NewDecoder(bytes.NewReader(v))
+	if useJSONNumber {
+		decoder.UseNumber()
+	}
+	if err := decoder.Decode(&jv); err != nil {
+		panic(err)
+	}
+	return jv
+}
+
+func ptrNumber(number document.Number) *document.Number {
+	return &number
+}
+
+func TestDecoder_DecodeJSONInterface(t *testing.T) {
+	t.Run("Object", func(t *testing.T) {
+		for name, tt := range sharedObjectTests {
+			t.Run(name, func(t *testing.T) {
+				testDecodeJSONInterface(t, tt)
+			})
+		}
+	})
+	t.Run("Array", func(t *testing.T) {
+		for name, tt := range sharedArrayTestCases {
+			t.Run(name, func(t *testing.T) {
+				testDecodeJSONInterface(t, tt)
+			})
+		}
+	})
+	t.Run("Number", func(t *testing.T) {
+		for name, tt := range sharedNumberTestCases {
+			t.Run(name, func(t *testing.T) {
+				testDecodeJSONInterface(t, tt)
+			})
+		}
+	})
+	t.Run("String", func(t *testing.T) {
+		for name, tt := range sharedStringTests {
+			t.Run(name, func(t *testing.T) {
+				testDecodeJSONInterface(t, tt)
+			})
+		}
+	})
+}
+
+func TestNoSerde(t *testing.T) {
+	type noSerde = document.NoSerde
+	type InvalidType struct {
+		FieldName string
+		noSerde
+	}
+	var v InvalidType
+
+	err := json.NewDecoder().DecodeJSONInterface(MustJSONUnmarshal([]byte(`{
+  "FieldName": "value"
+}`), true), &v)
+	if err == nil {
+		t.Fatalf("expect error, got nil")
+	}
+
+	type EmbedInvalidType struct {
+		FieldName string
+		TypeField InvalidType
+	}
+
+	var ev EmbedInvalidType
+
+	err = json.NewDecoder().DecodeJSONInterface(MustJSONUnmarshal([]byte(`{
+  "FieldName": "value",
+  "TypeField": {
+    "FieldName": "value"
+  }
+}`), true), &ev)
+	if err == nil {
+		t.Fatalf("expect error, got nil")
+	}
+}
+
+func TestNewDecoderUnsupportedTypes(t *testing.T) {
+	cases := []struct {
+		input             []byte
+		value             interface{}
+		disableJSONNumber bool
+	}{
```

Comment:
Would be helpful to also have a test case that the `Decoder` cannot unmarshal into a non-empty interface.